src/dbnode/client/write_state.go

```diff
@@ -260,3 +308,43 @@ type maybeHostWriteError struct {
 	// Error field is optionally set when there is actually an error.
 	err error
 }
+
+func newCountTowardsConsistency(
+	shardState shard.State,
+	leavingCountsIndividually bool,
+	leavingAndInitializingCountsAsPair bool,
+) countTowardsConsistency {
+
+	isAvailable := shardState == shard.Available
+	isLeaving := shardState == shard.Leaving
+	isInitializing := shardState == shard.Initializing
+
+	if isAvailable {
+		return available
+	}
+	if isLeaving && leavingCountsIndividually {
+		return shardLeavingCountsIndividually
+	}
+	if isLeaving && leavingAndInitializingCountsAsPair {
+		return shardLeavingCountsAsPair
+	}
+	if isInitializing && leavingAndInitializingCountsAsPair {
+		return shardInitializingCountsAsPair
+	}
+	if isLeaving {
+		return leaving
+	}
+	if isInitializing {
+		return initializing
+	}
+	return undefined
+}
+
+func findHost(hostSuccessList []ident.ID, hostID ident.ID) bool {
```

Comment:
nit: add a comment explaining why you chose to implement this as a for loop instead of a map.